dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1341,85 +1577,233 @@ RtpsUdpDataLink::send_ack_nacks(RtpsReaderMap::value_type& rr, bool finalFlag)
-                  sizeof(GuidPrefix_t));
-      ser << info_dst;
-      // Interoperability note: we used to insert INFO_REPLY submessage here, but
-      // testing indicated that other DDS implementations didn't accept it.
-      ser << acknack;
-      for (size_t i = 0; i < nack_frags.size(); ++i) {
-        nack_frags[i].readerId = rr.first.entityId;
-        nack_frags[i].writerId = wi->first.entityId;
-        ser << nack_frags[i]; // always 4-byte aligned
-      }
-
-      OPENDDS_SET(ACE_INET_Addr) addrs = get_addresses(rr.first, wi->first);
-      if (!addrs.empty()) {
-        send_strategy()->send_rtps_control(mb_acknack, addrs);
+  if (responses.empty()) {
+    return;
+  }
+
+  typedef OPENDDS_VECTOR(ResponseVec::iterator) ResponseIterVec;
+  typedef OPENDDS_MAP_CMP(RepoId, ResponseIterVec, GUID_tKeyLessThan) DestResponseMap;
+  typedef OPENDDS_MAP(AddrSet, DestResponseMap) AddrDestResponseMap;
+  AddrDestResponseMap adr_map;
+
+  {
+    ACE_GUARD(ACE_Thread_Mutex, g, lock_);
+    AddrSet addrs;
+    // Sort responses by address set and destination
+    for (ResponseVec::iterator it = responses.begin(); it != responses.end(); ++it) {
+      if (it->dst_guid_ == GUID_UNKNOWN) {
+        addrs = get_addresses_i(it->from_guid_); // This will overwrite, but addrs should always be empty here
       } else {
-        if (Transport_debug_level) {
-          const GuidConverter conv(wi->first);
-          ACE_ERROR((LM_ERROR,
-                     "(%P|%t) RtpsUdpDataLink::send_heartbeat_replies() - "
-                     "no locator for remote %C\n", OPENDDS_STRING(conv).c_str()));
+        accumulate_addresses(it->from_guid_, it->dst_guid_, addrs);
+      }
+      for (RepoIdSet::iterator it2 = it->to_guids_.begin(); it2 != it->to_guids_.end(); ++it2) {
+        accumulate_addresses(it->from_guid_, *it2, addrs);
+      }
+      if (addrs.empty()) {
+        continue;
+      }
+      if (std::memcmp(&(it->dst_guid_.guidPrefix), &GUIDPREFIX_UNKNOWN, sizeof(GuidPrefix_t)) != 0) {
+        RepoId dst;
+        memcpy(dst.guidPrefix, it->dst_guid_.guidPrefix, sizeof(dst.guidPrefix));
+        dst.entityId = ENTITYID_UNKNOWN;
+        adr_map[addrs][dst].push_back(it);
+      } else {
+        adr_map[addrs][GUID_UNKNOWN].push_back(it);
+      }
+      addrs.clear();
+    }
+  }
+
+  // Reusable INFO_DST
+  InfoDestinationSubmessage idst = {
+    {INFO_DST, FLAG_E, INFO_DST_SZ},
+    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+  };
+
+  // Build bundles
+  typedef OPENDDS_VECTOR(ResponseIterVec) ResponseIterVecVec;
+  ResponseIterVecVec response_bundles; // a vector of vectors of iterators pointing to responses
+  OPENDDS_VECTOR(AddrSet) response_bundle_addrs; // for a bundle's address set
+  OPENDDS_VECTOR(size_t) response_bundles_sizes; // for allocating the bundle's buffer
+  size_t size = 0, padding = 0, prev_size = 0, prev_padding = 0; // keep track of previous values in case we overshoot
```

Comment:
These are managed together as a group.  What would it looks like if it were its own struct with small inline methods for updating the `prev` and "resetting"?